stable/database/files/nuosm

```diff
@@ -387,6 +387,69 @@ function resolveLatestSource() {
   return 0
 }
 
+#=======================================
+# function - Looks for a previous archive and journal and imports it
+#
+function loadFromSnapshot() {
+  local RECREATE_ARCHIVES="false"
+
+  if [ -z "$(ls -A ${DB_DIR})" ]; then
+    local archives=$( find "/var/opt/nuodb/archive" -name info.json )
+    if [ -z "$archives" ] || [ $(echo "$archives" | wc -l) != 1 ]; then 
+      die 1 "Did not find exactly 1 archive: ${archives}"
+    fi
+
+    local ARCHIVE_PATH=$( dirname $archives )
+    log "Found snapshot at ${ARCHIVE_PATH}"
+
+    if [ -n "$BACKUP_ID" ]; then
+      if [ ! -e ${ARCHIVE_PATH}/backup.json ] || \
+          [ $BACKUP_ID != $(sed -e 's|.*"id":\s*"\?\([^\s"]\+\)"\?[,}].*|\1|' "$ARCHIVE_PATH/backup.json") ]; then
+        die 1 "Incorrect backup id in archive"
+      fi
+    fi
+
+    if [ "$SEPARATE_JOURNAL" == "true" ]; then
+      local RELATIVE_ARCHIVE=$(realpath --relative-base=/var/opt/nuodb/archive $ARCHIVE_PATH)
+      local EXPECTED_JOURNAL="/var/opt/nuodb/journal/${RELATIVE_ARCHIVE}"
+
+      if [ ! -d $EXPECTED_JOURNAL ]; then
+        die 1 "Did not find a journal snapshot at '$EXPECTED_JOURNAL'"
+      fi
+      
+      if [ -n "$BACKUP_ID" ]; then
+        if [ ! -e ${EXPECTED_JOURNAL}/backup.json ] || \
+            [ $BACKUP_ID != $(sed -e 's|.*"id":\s*"\?\([^\s"]\+\)"\?[,}].*|\1|' "$EXPECTED_JOURNAL/backup.json") ]; then
+          die 1 "Incorrect backup id in journal"
+        fi
+      fi
+
+      mv ${EXPECTED_JOURNAL}/* ${JOURNAL_DIR}
+
+    fi
+
+    # Don't move the archive until after we have done error checking.
+    # Once we move the archive, we are going to be in the `archive already exists` branch
+    # when kubernetes retries/restarts the pod.
+    mv ${ARCHIVE_PATH}/* ${DB_DIR}
+
+    local RECREATE_ARCHIVES="true"
+  else
+    log "archive already exists, not looking for snapshot archives"
+
+    if [ -e ${DB_DIR}/backup.json ] && [ -n "$BACKUP_ID" ] && \
+          [ $BACKUP_ID == $(sed -e 's|.*"id":\s*"\?\([^\s"]\+\)"\?[,}].*|\1|' "$DB_DIR/backup.json") ]; then
+      local RECREATE_ARCHIVES="true"
+    fi
```

Comment:
We check the `backup.json` in the journal directory above as a sanity check. Should we do the same here?

Also, we abort on mismatching backup ID above, but here we seem to disregard a mismatching backup ID.

Reply:
This is deliberate. We found a non-empty archive directory. It likely means that we this is just a regular SM restart and not a first boot.

If there is no backup id, that could mean that the archive was already recovered. We have, at least, deleted the info.json.

If there is a mismatched backup id, that could mean that we crashed while a snapshot was being taken.

Reply:
> if backup.json exist && $BACKUP_ID is not empty => verify the backup ID and fail on mismatch

That is a valid state of the SM, likely caused by a bug in our pre-snap logic.

For example:

1. A database is started from a backup with backup id `1`.
2. The SM starts up successfully and begins serving a database.
3. Some time later, a backup policy kicks in and triggers a snapshot.
4. `backup.json` is created with an id of `2`.
5. Due to a bug in the `fsfreeze` sidecar, the pod crashes.
6. Kubernetes sees a failed pod and restarts it.
7. `nuosm` starts with a Helm supplied $BACKUP_ID of `1` but `backup.json` says the id should be `2`.

The SM still has a valid archive and is registered with the Admin. It just needs to start and will be able to resume serving up the database.

Reply:
I feel like if the user is providing a `backup_id`, then we should verify it and fail on mismatch. If the user wants to disregard the validation, they can skip the `backup_id` altogether.

Reply:
So how do we differentiate the SM starting for the first time vs. the SM being restarted (and the SM being restarted while a snapshot is in progress)?

Reply:
We've discussed this offline and the issue that Kon is trying to solve here is that the `info.json` _shouldn't be removed_ on container restart in case the post-backup hook didn't remove the `backup.json` for some reason.
I think in case of a backup ID mismatch (which is an abnormal scenario), we should at least _log_.